### PR TITLE
[WIP] Rnn-T LM nbest rescoring

### PR DESCRIPTION
**Pull request by ezerhouni**
Code from https://github.com/k2-fsa/icefall/pull/365

Results :
```
For test-clean, WER of different settings are:
beam_4.0_max_contexts_8_max_states_32_num_paths_200_nbest_scale_0.5_temperature_1.0_ngram_lm_scale_0	1.99	best for test-clean
beam_4.0_max_contexts_8_max_states_32_num_paths_200_nbest_scale_0.5_temperature_1.0_ngram_lm_scale_0.01	1.99
beam_4.0_max_contexts_8_max_states_32_num_paths_200_nbest_scale_0.5_temperature_1.0_ngram_lm_scale_0.02	2.0
beam_4.0_max_contexts_8_max_states_32_num_paths_200_nbest_scale_0.5_temperature_1.0_ngram_lm_scale_0.1	2.0
beam_4.0_max_contexts_8_max_states_32_num_paths_200_nbest_scale_0.5_temperature_1.0_ngram_lm_scale_-0.02	2.02
beam_4.0_max_contexts_8_max_states_32_num_paths_200_nbest_scale_0.5_temperature_1.0_ngram_lm_scale_0.05	2.03
beam_4.0_max_contexts_8_max_states_32_num_paths_200_nbest_scale_0.5_temperature_1.0_ngram_lm_scale_-0.05	2.05
beam_4.0_max_contexts_8_max_states_32_num_paths_200_nbest_scale_0.5_temperature_1.0_ngram_lm_scale_-0.1	2.13
beam_4.0_max_contexts_8_max_states_32_num_paths_200_nbest_scale_0.5_temperature_1.0_ngram_lm_scale_0.3	2.16
beam_4.0_max_contexts_8_max_states_32_num_paths_200_nbest_scale_0.5_temperature_1.0_ngram_lm_scale_-0.2	2.31
beam_4.0_max_contexts_8_max_states_32_num_paths_200_nbest_scale_0.5_temperature_1.0_ngram_lm_scale_0.5	2.31
beam_4.0_max_contexts_8_max_states_32_num_paths_200_nbest_scale_0.5_temperature_1.0_ngram_lm_scale_0.8	2.42
beam_4.0_max_contexts_8_max_states_32_num_paths_200_nbest_scale_0.5_temperature_1.0_ngram_lm_scale_1.0	2.47
beam_4.0_max_contexts_8_max_states_32_num_paths_200_nbest_scale_0.5_temperature_1.0_ngram_lm_scale_1.5	2.56
beam_4.0_max_contexts_8_max_states_32_num_paths_200_nbest_scale_0.5_temperature_1.0_ngram_lm_scale_2.5	2.66
beam_4.0_max_contexts_8_max_states_32_num_paths_200_nbest_scale_0.5_temperature_1.0_ngram_lm_scale_3	2.68
beam_4.0_max_contexts_8_max_states_32_num_paths_200_nbest_scale_0.5_temperature_1.0_ngram_lm_scale_-0.5	2.99
```

```
For test-other, WER of different settings are:
beam_4.0_max_contexts_8_max_states_32_num_paths_200_nbest_scale_0.5_temperature_1.0_ngram_lm_scale_0.05	4.85	best for test-other
beam_4.0_max_contexts_8_max_states_32_num_paths_200_nbest_scale_0.5_temperature_1.0_ngram_lm_scale_0.1	4.85
beam_4.0_max_contexts_8_max_states_32_num_paths_200_nbest_scale_0.5_temperature_1.0_ngram_lm_scale_0.01	4.88
beam_4.0_max_contexts_8_max_states_32_num_paths_200_nbest_scale_0.5_temperature_1.0_ngram_lm_scale_0.02	4.88
beam_4.0_max_contexts_8_max_states_32_num_paths_200_nbest_scale_0.5_temperature_1.0_ngram_lm_scale_0	4.89
beam_4.0_max_contexts_8_max_states_32_num_paths_200_nbest_scale_0.5_temperature_1.0_ngram_lm_scale_-0.02	4.95
beam_4.0_max_contexts_8_max_states_32_num_paths_200_nbest_scale_0.5_temperature_1.0_ngram_lm_scale_-0.05	5.02
beam_4.0_max_contexts_8_max_states_32_num_paths_200_nbest_scale_0.5_temperature_1.0_ngram_lm_scale_0.3	5.02
beam_4.0_max_contexts_8_max_states_32_num_paths_200_nbest_scale_0.5_temperature_1.0_ngram_lm_scale_-0.1	5.16
beam_4.0_max_contexts_8_max_states_32_num_paths_200_nbest_scale_0.5_temperature_1.0_ngram_lm_scale_0.5	5.3
beam_4.0_max_contexts_8_max_states_32_num_paths_200_nbest_scale_0.5_temperature_1.0_ngram_lm_scale_0.8	5.53
beam_4.0_max_contexts_8_max_states_32_num_paths_200_nbest_scale_0.5_temperature_1.0_ngram_lm_scale_-0.2	5.6
beam_4.0_max_contexts_8_max_states_32_num_paths_200_nbest_scale_0.5_temperature_1.0_ngram_lm_scale_1.0	5.61
beam_4.0_max_contexts_8_max_states_32_num_paths_200_nbest_scale_0.5_temperature_1.0_ngram_lm_scale_1.5	5.76
beam_4.0_max_contexts_8_max_states_32_num_paths_200_nbest_scale_0.5_temperature_1.0_ngram_lm_scale_2.5	5.89
beam_4.0_max_contexts_8_max_states_32_num_paths_200_nbest_scale_0.5_temperature_1.0_ngram_lm_scale_3	5.94
beam_4.0_max_contexts_8_max_states_32_num_paths_200_nbest_scale_0.5_temperature_1.0_ngram_lm_scale_-0.5	6.63
```

### CODE REVIEW

**Comment by ezerhouni** [comment]
@csukuangfj I took your code and created a new PR otherwise the rebase was quite messy

**Comment by pkufool** [comment]
Cool! What's your baseline? How much can you benefit from LM rescoring.

**Comment by ezerhouni** [comment]
@pkufool Thank you !
The baseline should be for ```ngram_lm_scale_0```. It does better for test-other but equally for test-clean. I think a larger 4gram might help. I found one from fairseq I believe but it was a binary file so I could not plug it easily. I will try to see if I can find some time in the coming days to generate a larger one using different dataset and see if it helps.

**Comment by danpovey** [comment]
Great!  Even if the RNNLM rescoring does not give great improvemnents just yet, it's important to have the code available because I have a plan for a modified form of the RNN-T model training that should combine better with external LMs.

**Comment by csukuangfj** [comment]
The current PR is using n-gram, I think.

It would be great if you can use the RNN LM that you have trained in conformer ctc here; maybe in a separate PR.

**Comment by ezerhouni** [comment]
@csukuangfj Indeed, it is n-gram based for the moment. I can try to have a look into adding the RNN-LM rescoring into the RNN-T.
For the moment I am using the 4gram provided in the ```prepare.sh```. I was thinking to try regenerating the 4gram and adding gigaspeech corpus (to have a larger 4gram). I would rather not add any other external dataset (than gigaspeech) so that have a baseline with a "fix" dataset. What do you think ?

**Comment by csukuangfj** [comment]
> @csukuangfj Indeed, it is n-gram based for the moment. I can try to have a look into adding the RNN-LM rescoring into the RNN-T.
> 
> For the moment I am using the 4gram provided in the ```prepare.sh```. I was thinking to try regenerating the 4gram and adding gigaspeech corpus (to have a larger 4gram). I would rather not add any other external dataset (than gigaspeech) so that have a baseline with a "fix" dataset. What do you think ?

Thanks! That is also nice to have.

**Comment by ezerhouni** [comment]
@csukuangfj I opened a branch (on my repo) with RNN-LM rescoring. I have a bit of a weird results, but I didn't spend too much time debugging it. I will open a new PR/ update this one once I have more results to provide

**Comment by csukuangfj** [comment]
Thanks! Merging. The RNN LM stuff can be added in a separate PR.

**Comment by RuABraun** [comment]
How is this different from `fast_beam_search_nbest_LG` ? Is it that that would take `N` random paths from the lattice while this one actually gets the `N` best?

File: egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py [hunk 1]
```diff
@@ -1171,3 +1198,155 @@ def beam_search(
...
+    G: k2.Fsa,
+    sp: spm.SentencePieceProcessor,
+    word_table: k2.SymbolTable,
+    oov_word: str = "<UNK>",
+    use_double_scores: bool = True,
+    nbest_scale: float = 0.5,
+    temperature: float = 1.0,
+) -> Dict[str, List[List[int]]]:
+    """It limits the maximum number of symbols per frame to 1.
+    A lattice is first obtained using modified beam search, and then
```
Comment (danpovey) [inline egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py:1222]: I don't think this comment is accurate?  It doesn't mention anything about n-best rescoring.  Please make sure all args are well documented.

**Comment by danpovey** [comment]
> How is this different from `fast_beam_search_nbest_LG` ? Is it that that would take `N` weighted random paths from the lattice while this one actually gets the `N` best?

This one will also be pseudo-random, I dpon't know the difference, will wait for one of the others to respond.

**Comment by csukuangfj** [comment]
fast_beam_search_nbest_LG uses an LG graph to replace the trivial graph while generating the lattice. After generating the lattice, it selects n paths form the lattice; outputs the one with the highest score.

While this PR uses a trivial graph to generate the lattice. Also, it uses the given G to rescore the selected n paths.

**Comment by danpovey** [comment]
Ah OK, perhaps something like that could be put in the comment at some point?

**Comment by csukuangfj** [comment]
> Ah OK, perhaps something like that could be put in the comment at some point?

Yes, will do.

**Comment by ezerhouni** [comment]
@csukuangfj I can update the comments of both functions in https://github.com/k2-fsa/icefall/pull/475

**Comment by csukuangfj** [comment]
> @csukuangfj I can update the comments of both functions in #475

Thanks! Please go ahead.

**Pull request merged**